### PR TITLE
Fix Process.run uses in compiler to handle exec failure

### LINKED ISSUE #9789: Confusing assertion error when building docs without git installed
Building 0.35.1 in a clean environment (FreeBSD packaging) would fail with a confusing

```
bin/crystal docs src/docs_main.cr  --project-name=Crystal --project-version=0.35.1 --source-refname=
bin/crystal: crystal: not found
Using compiled compiler at .build/crystal
Unhandled exception: Crystal::Compiler#program cannot be nil (NilAssertionError)
  from __crystal_realloc
  from __crystal_sigfault_handler
  from __crystal_sigfault_handler
  from __crystal_main
  from main
  from _start
```

turns out, right after it tries to execute git:

```
12598: execve("/usr/local/sbin/git",0x806dd74c0,0x7fffffffda60) ERR#2 'No such file or directory'
12598: execve("/usr/local/bin/git",0x806dd74c0,0x7fffffffda60) ERR#2 'No such file or directory'
12598: execve("/usr/sbin/git",0x806dd74c0,0x7fffffffda60) ERR#2 'No such file or directory'
12598: execve("/usr/bin/git",0x806dd74c0,0x7fffffffda60) ERR#2 'No such file or directory'
12598: execve("/sbin/git",0x806dd74c0,0x7fffffffda60) ERR#2 'No such file or directory'
12598: execve("/bin/git",0x806dd74c0,0x7fffffffda60) ERR#2 'No such file or directory'
12598: write(15,"\^A",1)                         = 1 (0x1)
 7355: kevent(15,{ 14,EVFILT_READ,EV_ADD,0,0,0x30303 8,EVFILT_READ,EV_ADD,0,0,0x30303 11,EVFILT_READ,EV_ADD,0,0,0x30303 },3,{ 14,EVFILT_READ,0x0,0,0x1,0x30303 },64,{ 4.999951000 }) = 1 (0x1)
12598: write(15,"\^B\0\0\0",4)                   = 4 (0x4)
 7355: read(14,"\^A\^B\0\0\0",8192)              = 5 (0x5)
 7355: mmap(0x0,28672,PROT_READ|PROT_WRITE,MAP_PRIVATE|MAP_ANON,-1,0x0) = 34508636160 (0x808e00000)
 7355: fstatat(AT_FDCWD,"/usr/share/nls/C/libc.cat",0x7fffffffbcb0,0x0) ERR#2 'No such file or directory'
 7355: fstatat(AT_FDCWD,"/usr/share/nls/libc/C",0x7fffffffbcb0,0x0) ERR#2 'No such file or directory'
 7355: fstatat(AT_FDCWD,"/usr/local/share/nls/C/libc.cat",0x7fffffffbcb0,0x0) ERR#2 'No such file or directory'
 7355: fstatat(AT_FDCWD,"/usr/local/share/nls/libc/C",0x7fffffffbcb0,0x0) ERR#2 'No such file or directory'
 7355: close(14)                                 = 0 (0x0)
Unhandled exception: Crystal::Compiler#program cannot be nil (NilAssertionError)
```

and installing git fixes it.

Would be nice to make git not required (probably there's already a way to provide the git rev or whatever it wants?) but at least please make a **good error message**!

### PR DESCRIPTION
**Pull request by straight-shoota**
Fixes more uses of `Process.run` to handle exec error (see #9867 and #9789). Also refactors the git code used in ProjectInfo to a common location for reusability.

See https://github.com/crystal-lang/crystal/pull/9867#issuecomment-723262588

<del/>Depends on #9867</del> <ins>includes #9867</ins>
Fixes #9789

### CODE REVIEW

**Comment by straight-shoota** [comment]
This could also be merged instead of #9867

File: src/compiler/crystal/tools/doc/project_info.cr [hunk 1]
```diff
@@ -133,44 +130,35 @@ module Crystal::Doc
 
     def self.git_clean?
       # Use git to determine if index and working directory are clean
-      io = IO::Memory.new
-      status = Process.run("git", ["status", "--porcelain"], output: io)
+      capture = Crystal::Git.git_capture(["status", "--porcelain"]) { return }
+
       # If clean, output of `git status --porcelain` is empty. Still need to check
```
Comment (bcardiff) [inline src/compiler/crystal/tools/doc/project_info.cr:135]: I guess the comment should be placed within the `{ return }` block, right?

File: src/compiler/crystal/tools/init.cr [hunk 2]
```diff
@@ -96,22 +97,16 @@ module Crystal
...
 
-    private def self.git_config(key)
-      String.build do |io|
-        Process.run("git", ["config", "--get", key], output: io)
-      end.strip.presence
-    end
-
     def self.fetch_author
-      git_config("user.name") || "your-name-here"
+      Crystal.git_config("user.name") || "your-name-here"
```
Comment (bcardiff) [inline src/compiler/crystal/tools/init.cr:101]:
```suggestion
      Crystal::Git.git_config("user.name") || "your-name-here"
```

right?

File: src/compiler/crystal/tools/doc/project_info.cr (hunk 1 above)
Comment (straight-shoota) [reply src/compiler/crystal/tools/doc/project_info.cr:135]: No, why?

File: src/compiler/crystal/tools/init.cr (hunk 2 above)
Comment (straight-shoota) [reply src/compiler/crystal/tools/init.cr:101]: Ooops. Forgot to push that part. Fixed.

File: src/compiler/crystal/tools/doc/project_info.cr (hunk 1 above)
Comment (bcardiff) [reply src/compiler/crystal/tools/doc/project_info.cr:135]: git_capture checks the status code. The `return unless status.success?` is gone.

Comment (straight-shoota) [reply src/compiler/crystal/tools/doc/project_info.cr:135]: Oh, sry I was confused. You're right.

**Review by bcardiff — APPROVED** [review]

**Comment by bcardiff** [comment]
@straight-shoota I think we can go with this PR as is instead of #9867, but I would need you to remove the Draft status.

File: src/compiler/crystal/tools/git.cr [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+    status
+  rescue IO::Error
+    yield
+  end
+
+  def self.git_capture(args)
+    String.build do |io|
+      git_command(args, output: io) { yield }
+    end
+  end
```
Comment (Sija) [inline src/compiler/crystal/tools/git.cr:18]: @straight-shoota Why `yield` if use can just use nilable return type to signify failure?

Comment (straight-shoota) [reply src/compiler/crystal/tools/git.cr:18]: Oh, I guess that's just left over from an initial implementation. I agree that returning `nil` seems more reasonable.

File: src/compiler/crystal/tools/git.cr [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+
+  # Tries to run git command with args.
+  # Yields block if exec fails or process status is not success.
+  def self.git_command(args, output : Process::Stdio = Process::Redirect::Close)
+    status = Process.run(executable, args, output: output)
+    yield unless status.success?
+    status
+  rescue IO::Error
+    yield
+  end
```
Comment (Sija) [inline src/compiler/crystal/tools/git.cr:12]: ditto

File: src/compiler/crystal/tools/git.cr [hunk 5]
```diff
@@ -0,0 +1,22 @@
+module Crystal::Git
+  class_property executable = "git"
+
+  # Tries to run git command with args.
+  # Yields block if exec fails or process status is not success.
+  def self.git_command(args, output : Process::Stdio = Process::Redirect::Close)
+    status = Process.run(executable, args, output: output)
+    return status.success?
```
Comment (Sija) [inline src/compiler/crystal/tools/git.cr:8]:
```suggestion
    status.success?
```

Comment (straight-shoota) [reply src/compiler/crystal/tools/git.cr:8]: Yes, return is implicit, but I'd prefer to be more explicit.

**Review by bcardiff — APPROVED** [review]

**Comment by oprypin** [comment]
This PR never passed Windows CI, and that was not a fluke.

https://github.com/crystal-lang/crystal/actions?query=workflow%3A%22Windows+CI%22+branch%3Amaster

image

We need quick action; unfortunately Windows CI is broken in *three* different way on _master_ right now.

**Comment by bcardiff** [comment]
I can merge #9912 right away. After that the specs of this PR can be worked. Does that sound good?

**Comment by straight-shoota** [comment]
@oprypin Sorry, I didn't realize the failure was an acutal bug in this PR. Let's make CI reliable again =)

File: spec/compiler/crystal/tools/doc/project_info_spec.cr [hunk 6]
```diff
@@ -33,7 +33,17 @@ describe Crystal::Doc::ProjectInfo do
...
 
-      it "no git" do
+      it "git missing" do
+        Crystal::Git.executable = "git-missing-executable"
+
+        assert_with_defaults(ProjectInfo.new(nil, nil), ProjectInfo.new("foo", "1.0", refname: nil))
+        assert_with_defaults(ProjectInfo.new("bar", "2.0"), ProjectInfo.new("bar", "2.0", refname: nil))
+        assert_with_defaults(ProjectInfo.new(nil, "2.0"), ProjectInfo.new("foo", "2.0", refname: nil))
+      ensure
+        Crystal::Git.executable = "git"
```
Comment (oprypin) [inline spec/compiler/crystal/tools/doc/project_info_spec.cr:43]: I confirm that the specs get stuck immediately after exiting this spec. But I can't figure out where. Both the `around_each` and `before_each` above also finish, and *then* somehow spec gets stuck, before entering the next spec.

Comment (straight-shoota) [reply spec/compiler/crystal/tools/doc/project_info_spec.cr:43]: We should probably just add a quick fix to disable this spec on win32 to get CI working.

Comment (oprypin) [reply spec/compiler/crystal/tools/doc/project_info_spec.cr:43]: I think I understand now. Of course, the bug in entirely in Crystal's support of Windows.

This leaves a dangling fiber that's still trying to write to the non-launched process. While background fibers getting permanently stuck is a known (horrible) bug, perhaps there's another bug that forgets to close something if the process never actually launched.

Repro in `cmd`:

```cmd
crystal eval "p! Process.run(%(missing), output: IO::Memory.new) rescue nil; p! Process.run(%(git), output: IO::Memory.new)"
```

Comment (oprypin) [reply spec/compiler/crystal/tools/doc/project_info_spec.cr:43]: @straight-shoota Yes, and looks like that'll even be the somewhat long-term fix. #9913

**Pull request merged**